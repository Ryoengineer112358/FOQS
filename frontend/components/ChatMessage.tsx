import React from 'react'
import SendMessage from './SendMessage'
import { StudentQuestion, TutorAnswer, StudentComment } from '@/types'

type MessageStudent = StudentComment & { className: string }
type MessageTutor = TutorAnswer & { className: string }
type TimelineItem = MessageStudent | MessageTutor

type Props = {
  middleware: string
  question: StudentQuestion
  sendFunction: Function
}

function ChatMessage(props: Props) {
  return (
    <>
      {(() => {
        const messageTutor = props.question.tutor_answers.map((t) => {
          return {
            ...t,
            className: props.middleware === 'tutor' ? 'sent' : 'received',
          }
        })
        const messageStudent = props.question.student_comments.map((s) => {
          return {
            ...s,
            className: props.middleware === 'student' ? 'sent' : 'received',
          }
        })
        console.log(messageTutor)
        const messages = messageTutor
          .concat(messageStudent)
          .sort(
            (a: TimelineItem, b: TimelineItem) =>
              a.created_at.getTime() - b.created_at.getTime(),
          )
        return (
          <div className='msgs'>
            {props.question.image_urls &&
              props.question.image_urls.map((url, index) => (
                <div key={index} className='image'>
                  <img
                    src={`${process.env.NEXT_PUBLIC_BACKEND_URL}/storage/${url}`}
                    style={{ width: '100%' }}
                    alt='生徒が質問をした時に送った画像'
                  />
                </div>
              ))}
            {props.question.text ? (
              <div>
                <div
                  className={`msg ${
                    props.middleware === 'student' ? 'sent' : 'received'
                  }`}
                >
                  <p>{props.question.text}</p>
                </div>
              </div>
            ) : (
              <></>
            )}
            {messages.map((message) => (
              <div key={message.id}>
                <div className={`msg ${message.className}`}>
                  <p>{message.text}</p>
                </div>
              </div>
            ))}
          </div>
        )
      })()}
      {props.question.student_id !== 0 &&
        (props.question.closed_at ? (
          <></>
        ) : (
          <SendMessage
            sendFunction={props.sendFunction}
            sender_role={props.middleware}
          />
        ))}
    </>
  )
}

export default ChatMessage
