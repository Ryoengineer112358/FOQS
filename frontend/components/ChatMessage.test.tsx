import React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import ChatMessage from './ChatMessage'

test('ChatMessage component tests', () => {
  const question = {
    id: 1,
    created_at: new Date('2023-01-01'),
    updated_at: new Date('2023-01-01'),
    tutor_answers: [
      { id: 1, text: 'Tutor Answer 1', created_at: new Date('2023-01-01'), tutor_id: 1, student_question_id: 1, updated_at: new Date('2023-01-01') },
      { id: 2, text: 'Tutor Answer 2', created_at: new Date('2023-01-02'), tutor_id: 1, student_question_id: 1, updated_at: new Date('2023-01-02') },
    ],
    student_comments: [
      { id: 3, text: 'Student Comment', created_at: new Date('2023-01-03'), tutor_id: 1, student_question_id: 1, updated_at: new Date('2023-01-03') },
    ],
    image_urls: ['image1.jpg'],
    text: 'Question Text',
    student_id: 1,
    //closed_at: null,
  }
  
  
  const middleware = 'tutor'
  const sendFunction = jest.fn()

  render(
    <ChatMessage
      middleware={middleware}
      question={question}
      sendFunction={sendFunction}
    />,
  )

  // メッセージ入力フォームが表示されていることを確認
  const sendMessageElement = screen.getByRole('button', { name: /send/i })
  expect(sendMessageElement).toBeInTheDocument()

  // 質問のテキストが表示されていることを確認
  const questionTextElement = screen.getByText(/Question Text/i)
  expect(questionTextElement).toBeInTheDocument()

  // 画像が表示されていることを確認
  const imageElement = screen.getByRole('img')
  expect(imageElement).toHaveAttribute(
    'src',
    `${process.env.NEXT_PUBLIC_BACKEND_URL}/storage/image1.jpg`,
  )

  // メッセージが正しい順序で表示されていることを確認
  const messages = screen.getAllByRole('article') // <p>タグの代わりに<article>タグを使用していると仮定
  expect(messages[0]).toHaveTextContent('Tutor Answer 1')
  expect(messages[1]).toHaveTextContent('Tutor Answer 2')
  expect(messages[2]).toHaveTextContent('Student Comment')
})
