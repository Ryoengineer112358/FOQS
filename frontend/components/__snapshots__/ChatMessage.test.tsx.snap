// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessage component tests 1`] = `
<div
  className="msgs"
>
  <div
    className="image"
  >
    <img
      alt="生徒が質問をした時に送った画像"
      src="undefined/storage/image1.jpg"
      style={
        {
          "width": "100%",
        }
      }
    />
  </div>
  <div>
    <div
      className="msg received"
    >
      <p>
        Question Text
      </p>
    </div>
  </div>
  <div>
    <div
      className="msg sent"
    >
      <p>
        Tutor Answer 1
      </p>
    </div>
  </div>
  <div>
    <div
      className="msg sent"
    >
      <p>
        Tutor Answer 2
      </p>
    </div>
  </div>
  <div>
    <div
      className="msg received"
    >
      <p>
        Student Comment
      </p>
    </div>
  </div>
</div>
`;
